### DISCUSSION TITLE
MIT 6.006: Introduction to Algorithms - CS自学指南

### DISCUSSION THREAD

**Discussion by giscus** [PKUFlyingPig/cs-self-learning#395] [Announcements]
# MIT 6.006: Introduction to Algorithms - CS自学指南

CS自学指南

https://csdiy.wiki/%E6%95%B0%E6%8D%AE%E7%BB%93%E6%9E%84%E4%B8%8E%E7%AE%97%E6%B3%95/6.006/

**Comment by xlxdxy** [comment]
另外有一个 2020 年的版本，[Introduction to Algorithms | spring 2020](https://ocw.mit.edu/courses/6-006-introduction-to-algorithms-spring-2020/)

**Reply by taylover2016** [reply]
关于课程版本，目前b站可搜集到的一共有三种：

最老的是 [Fall 2005 的版本](https://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-046j-introduction-to-algorithms-sma-5503-fall-2005/), 授课老师是 Charles Leiserson （算法导论的作者之一）和 Erik Demaine. 

其次是 [Fall 2011 的版本](https://ocw.mit.edu/courses/6-006-introduction-to-algorithms-fall-2011/), 授课老师是 Srini Devadas 和 Erik Demaine. 

最新版是 [Spring 2020 的版本](https://ocw.mit.edu/courses/6-006-introduction-to-algorithms-spring-2020/), 授课老师是 Erik Demaine, Jason Ku 和 Justin Solomon.

三者的关系大概是：Fall 2005 的 6.046 = Fall 2011 的 6.006 + Spring 2015 的 6.046 = Spring 2020 的 6.006, 不过个人感觉第二种搭配讲解最为详细。2020 的版本删减简化较多，而 2005 的版本难度相对较大。

**Reply by Gunale0926** [reply]
2020 6.006 = 2011 6.006 +  2015 6.046？不太可能吧。

**Reply by HlinForest** [reply]
Demaine在主页上说新版的6.006 replacing half of old 6.046

**Comment by AlexanderWang888** [comment]
建了个群 630708930

**Comment by rinevard** [comment]
在fall 2011的版本中，python版本是2.x而非我们现在常用的3.x，因此，会出现代码在3.x版本中报错的情况。
一个很明显的区别就是，在2.x中，print是一个语句而不是一个函数：

print "Hello, World!"

而在3.x中，print被转换为了一个函数：

print("Hello, World!")


为此，在windows系统中，可以在命令行中使用以下命令把2.x的python代码自动转换为3.x的代码：

python -m lib2to3 -w your_old_script.py

**Reply by kirkzhangtech** [reply]
我创建了个群
259386722

**Comment by kirkzhangtech** [comment]
我创建了个群，方便大家公开课内容
259386722

**Comment by MAXUPUPUP** [comment]
# 梦开始的地方！

**Comment by fu829** [comment]
想請問一下，用py實踐，或是用Cpp實踐，對未來就業會有什麼樣的影響嗎?

**Reply by Ydefrr** [reply]
个人觉得影响不大，因为我们学的数据结构与算法而不是编程语言，可以用自己熟悉的语言实现

**Comment by lbj031221** [comment]
有没有C++的数据结构与算法课程推荐

**Reply by tanxy2024** [reply]
求问 有找到吗

**Reply by axlzzd226** [reply]
可以在“一个仅供参考的CS学习规划”这里页面的算法与数据结构章节下找下，有Cpp的课。

**Reply by ManfredShao** [reply]
Stanford CS106B/X

**Comment by minwang-ai** [comment]
这个课有什么证书认证吗？没搜到任何

**Comment by Changsheng0727** [comment]
非常好的课程，使我的数据结构旋转

**Reply by ddddddavid66** [reply]
跟CS61B比哪个更难？ 我学完61B了求问

**Comment by kawayee** [comment]
我是为了看6.046来刷6.006(2020)的，我已经刷过cs61b、algs4。
6.006不适合作为没有数据结构基础的人的第一门课程，cs61b侧重apply problem，algs4侧重mechanical problem，6.006侧重modify problem，有点像“现在你已经学会1+1=2了，让我们算一道微积分吧”。
6.006非常强调知识的结构化，比如：他会先定义2个接口sequence interface、set interface，然后用各种数据结构来实现这两个接口，同时分析各种数据结构的优缺点。
课堂风格：非常像算法导论，充满了大量的数学符号和证明。
课堂结构：背景+定义术语+引理证明+算法介绍+算法正确性和复杂度分析。
resource：note浓缩lecture内容，recitation拓展lecture内容，problem session经典题型，problem set练习题，题目侧重于算法设计，不要求写代码，但要求数学证明。
一个省时间小技巧：把pdf喂给gpt，让他逐条解释。
